src/mimick_utils.hpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MIMICK_UTILS_HPP_
+#define MIMICK_UTILS_HPP_
+
+#include <limits>
+#include <memory>
+
+#include "mimick/mimick.h"
+
+template<typename Fn, Fn fn, class C, typename ... A>
+typename std::result_of<Fn(C, A...)>::type
+mmk_wrapper(A... a)
+{
+  struct mmk_stub * stub = mmk_ctx();
+  assert(nullptr != stub);
+  C * instance = static_cast<C *>(mmk_stub_context(stub));
+  assert(nullptr != instance);
+  return (instance->*fn)(a ...);
+}
+
+#define mmk_stub_create_wrapped(name, func, inst, ...) \
+  mmk_stub_create( \
+    name, \
+    (mmk_fn) & mmk_wrapper< \
+      decltype(&std::remove_reference<decltype(*inst)>::type::func), \
+      &std::remove_reference<decltype(*inst)>::type::func, \
+      std::remove_reference<decltype(*inst)>::type, \
+      __VA_ARGS__>, \
+    inst);
```

Comment:
@cottsay meta: some of this machinery is already available at https://github.com/brawner/test_mocking_utils/pull/1. This one's neat though, perhaps we can mix and match.

Reply:
I don't have any plans of making `test_mocking_utils` official. So if you want to make use of the ideas and steal code, feel free to incorporate them here.